### PR TITLE
fix: Deleting Postgres CR should delete PostgresUsers

### LINKED ISSUE #44: Can't delete PostgresUser CR after deleting Postgres CR
**Describe the bug**
The bug is following:
If you have deleted the Postgres CR linked to PostgresUser CR, you can't properly delete a PostgresUser anymore. The operator will loop over an error in case `dropOnDelete` is set to `true` or at restart, it will crash.

**To Reproduce**
1. Create a Postgres CR
```yaml
apiVersion: db.movetokube.com/v1alpha1
kind: Postgres
metadata:
  name: my-db
  namespace: app
spec:
  database: test-db # Name of database created in PostgreSQL
  dropOnDelete: true # Set to true if you want the operator to drop the database and role when this CR is deleted
```
2. Create a PostgresUser CR
```yaml
apiVersion: db.movetokube.com/v1alpha1
kind: PostgresUser
metadata:
  name: my-db-user
  namespace: app
spec:
  role: username
  database: my-db # This references the Postgres CR
  secretName: my-secret
  privileges: OWNER # Can be OWNER/READ/WRITE
```
3. Delete Postgres CR
4. Delete PostgresUser CR
5. See error

**Expected behavior**
For me, an expected behaviour should be to delete all PostgresUser linked to a Postgres CR when this one is deleted.

**Screenshots**
None

**Version and platform (please complete the following information):**
 - Platform [e.g. Linux, Mac, Windows, ...]: Linux
 - Arch [e.g. amd64, ...]: amd64
 - Version [e.g. 22]: 0.4.2

**Additional context**
None

### PR DESCRIPTION
**Pull request by oxyno-zeta**
Will fix #44 by deleting all PostgresUser CR linked to a Postgres CR that was deleted.

### CODE REVIEW

**Comment by oxyno-zeta** [comment]
I just saw that I broke all tests... I will fix them.

**Comment by oxyno-zeta** [comment]
Hello @arnarg , Ok I will work on this :) . I will edit that pull request. Thanks for your first review.

**Comment by oxyno-zeta** [comment]
Hello @arnarg , 

I've edited the code to manage Cascading deletion and foreground is not required. 
In order to delete a PostgresUser CR linked to a non existing Postgres CR, I've used the default database.

Is it ok for you ?

**Comment by arnarg** [comment]
> It tries to delete the Postgres CR and after the deletion, delete all children.

To me it sounds like you're describing background cascading deletion
```
In background cascading deletion, Kubernetes deletes the owner object immediately and the garbage collector then deletes the dependents in the background.
```

The way I understand the documentation we have to check if there is a finalizer `foregroundDeletion` present in the Postgres CR and if so, requeue until it is not present before we can drop the roles and database.

It says:

>In foreground cascading deletion, the root object first enters a “deletion in progress” state. In the “deletion in progress” state, the following things are true:
>
> - The object is still visible via the REST API
> - The object’s deletionTimestamp is set
> - The object’s metadata.finalizers contains the value “foregroundDeletion”.

We are just checking if the deletionTimestamp is set and therefor we are not waiting for the "owned" PostgresUser CR to be deleted first.

**Comment by oxyno-zeta** [comment]
In fact, the kubectl command is blocked but until the Postgres CR is deleted, nothing more happened.
The Postgres CR is always requeued and nothing change. 

This is done asynchronously, it is better in a certain way no ?

File: pkg/controller/postgresuser/postgresuser_controller.go [hunk 1]
```diff
@@ -137,15 +150,25 @@ func (r *ReconcilePostgresUser) Reconcile(request reconcile.Request) (reconcile.
...
+
+	// Add owner to Postgres CR
+	// Get database object
+	database, err := r.getPostgresCR(instance)
+	if err != nil {
+		return r.requeue(instance, errors.NewInternalError(err))
+	}
+
+	// Add owner
+	err = controllerutil.SetControllerReference(database, instance, r.scheme)
```
Comment (hitman99) [inline pkg/controller/postgresuser/postgresuser_controller.go]: Why is this needed?  Owner is already set with [`addOwnerRef`](https://github.com/movetokube/postgres-operator/blob/64913c4e4acaa67a99cbcb90f6660a62be31d4ee/pkg/controller/postgresuser/postgresuser_controller.go#L220), isn't this doing the same thing again?

File: pkg/controller/postgresuser/postgresuser_controller.go [hunk 2]
```diff
@@ -137,15 +150,25 @@ func (r *ReconcilePostgresUser) Reconcile(request reconcile.Request) (reconcile.
 	}
 
 	// Creation logic
+
+	// Add owner to Postgres CR
+	// Get database object
+	database, err := r.getPostgresCR(instance)
```
Comment (hitman99) [inline pkg/controller/postgresuser/postgresuser_controller.go]: Can you elaborate a bit more on why this is moved from the `if` block below and how is it different?

**Comment by hitman99** [comment]
Also, do you mind merging changes from master? I fixed the github action to run tests on pull requests too

**Comment by oxyno-zeta** [comment]
Yes I can rebase from master. I will do it tomorrow and check your reviews.

**Comment by oxyno-zeta** [comment]
I've updated the code with your reviews points.
I've also rebased from master to add Pull Request builds.

**Comment by hitman99** [comment]
> In fact, the method called `addOwnerRef` didn't work because `OwnerReferences` is nil at the beginning and append to nil didn't do anything.

Are you sure about this? I think it works just fine, I did e2e tests for this specifically and as I remember it has worked correctly. I need to double check this in my setup to see it fail

**Comment by arnarg** [comment]
What's happening is that the PostgresUser object is never updated after adding the OwnerRef. If you do `r.Client.Update` in the function explicitly it works but we have to check if the same owner ref is already present, otherwise Kubernetes complains.

**Comment by oxyno-zeta** [comment]
@arnarg : I don't have any errors with the update even if "nothing" change. Everything is fine on minikube 1.15.

@hitman99 : When I've tested it nothing happened and double checked with debug. That's why I've updated the code.

**Comment by arnarg** [comment]
> I don't have any errors with the update even if "nothing" change. Everything is fine on minikube 1.15.

Yeah `controllerutil.SetControllerReference` takes care of not adding duplicates. I was talking about the code prior to your changes where the PostgresUser wasn't being updated after adding a owner ref and if that was added it would add duplicates in later reconcile loops.

**Comment by hitman99** [comment]
I had some time yesterday to test this and ir loooks good, except one thing. Every time I delete `Postgres` CR I get a not found error from postgres controller after it drops the database. It deletes both `Postgres` and linked `PostgresUser` regardless. Can you reproduce it?

**Comment by oxyno-zeta** [comment]
@hitman99 Yes I can reproduce it. This comes from the `defer` patch in the `Postgres` controller. This happened when the delete appeared but I think it isn't related to this fix.

**Comment by hitman99** [comment]
@oxyno-zeta yeah, that makes sense

**Comment by arnarg** [comment]
The deferred patch patches status separately after patching the rest (because it's a subresource). So when deleting it first removes the finalizer with a patch and then tries to patch the status but Kubernetes has already garbage collected the object. I'm guessing that's what is happening here.

It's benign but annoying.

**Comment by arnarg** [comment]
I guess we could patch the status first.

**Comment by hitman99** [comment]
I'm going to merge this

**Pull request merged**